server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -85,9 +109,37 @@ private FieldFetcher(List<FieldContext> fieldContexts) {
                 documentFields.put(field, new DocumentField(field, parsedValues));
             }
         }
+        if (includeUnmapped) {
+            Map<String, Object> unmappedFieldsToAdd = filter.apply(sourceLookup.loadSourceIfNeeded());
+            collectLeafValues(unmappedFieldsToAdd, documentFields);
+        }
         return documentFields;
     }
 
+    static void collectLeafValues(Map<String, Object> map, Map<String, DocumentField> documentFields) {
+        collectAllPaths("", map, documentFields);
+    }
+
+    private static void collectAllPaths(String prefix, Map<String, Object> source, Map<String, DocumentField> documentFields) {
+        for (String key : source.keySet()) {
+            Object value = source.get(key);
+            String currentPath = prefix + key;
+            if (value instanceof Map) {
+                collectAllPaths(currentPath + ".", (Map<String, Object>) value, documentFields);
+            } else {
+                DocumentField f;
```

Comment:
I think this might miss lists of objects? Looking at `XContentMapValues#filter`, there's a special recursive case for lists.

Reply:
I will take a look but though I have a test for this (essentially array values?)

Reply:
Ah, do you mean walking e.g. Lists of objects to their leaf values? e.g.
```
{
    "foo" : [ { "f1" : "value1" }, { "f2" : "value2" } ]
}
```
How does the current "fields" lookup work for this, e.g. what does the path look like for the "f1" value= "foo.1.f1" ? Maybe I'm just confused here.

Reply:
Sorry for the confusion, I read the logic too quickly before. We won't miss anything, but the result structure could be surprising.

For mapped fields, the 'fields' option always flattens arrays of objects. For example, given a document like

```
"object": [{ "field": "value1"}, {"field": "value2" }]
```

a request for `"fields": ["object.field1"]` will return

```
"fields": { "object.field": ["value1", "value2"]}
```

With this logic, for unmapped fields we'll return 

```
"fields": { "object": [{"field": "value1"}, {"field": "value2"} ]}
```

Perhaps this behavior is inconsistent, I think we'll still want to flatten arrays of objects (especially if we already flatten objects when they're not in array?) I'm definitely up for discussing this though, it's a question we noted on the original issue.

Reply:
I understand now and totally agree. I think I ran into this working on improvements on Friday anyway, should be part of my next update here.